conda/file_permissions.py

```diff
@@ -0,0 +1,128 @@
+
+
+class FilePermissions(object):
+
+    def __init__(self, prefix):
+        self.prefix = prefix
+
+    def _compose_file_structure(self, files):
+        file_structure = {"root": []}
+        for f in files:
+            file_elements = f.split("/")
+            abs_depth = len(file_elements)
+            for el in file_elements:
+                depth = file_elements.index(el)
+                el_type = FILE if depth+1 == abs_depth else DIR
+                path = "/".join(file_elements[0:depth])
+                child_path = "/".join(file_elements[0:depth + 1])
+                child_node = (child_path, el_type)
+                if depth == 0:
+                    root_el = (el, el_type)
+                    if root_el not in file_structure.get("root"):
+                        file_structure["root"].append((el, el_type))
+                elif depth < abs_depth:
+                    existing_child_nodes = file_structure.get(path)
+                    if path in file_structure.keys() and child_node not in existing_child_nodes:
+                        existing_child_nodes.append(child_node)
+                    elif path not in file_structure.keys():
+                        file_structure[path] = [child_node]
+        return file_structure
+
+    def _check_file(self, path, unlink_files):
+        is_being_unlinked = path in unlink_files
+        if not os.path.exists(path) or is_being_unlinked:
+            self.check_write_permission(dirname(path))
+        else:
+            raise CondaFileIOError(path, "File already exists, cannot link")
+
+    def _check_dir(self, dst, structured_link_files, unlink_files):
+        path = join(self.prefix, dst)
+        if os.path.exists(path):
+            self._check_files_permissions(structured_link_files, dst, unlink_files)
+        else:
+            self.check_write_permission(dirname(path))
+
+    def _check_files_permissions(self, structured_link_files, structured_files_root, unlink_files):
+        paths = structured_link_files.get(structured_files_root)
+        for path in paths:
+            if path[1] == DIR:
+                self._check_dir(path[0], structured_link_files, unlink_files)
+            elif path[1] == FILE:
+                self._check_file(join(self.prefix, path[0]), unlink_files)
+
+    def _can_write(self, path):
+        # On windows, "modify" permission is required on directories and files. This means that the
+        # user is able to create, rename, write to and delete files or directories
+        if not os.path.exists(path):
+            return False
+        if os.path.isdir(path):
+            test_file = join(path, "tmp-check-write-perms")
+            try:
+                open(test_file, 'a').close()
+                os.remove(test_file)
+            except IOError:
+                return False
+        elif os.path.isfile(path):
+            try:
+                renamed_path = "%s-test-rename" % path
+                os.rename(path, renamed_path)
```

Comment:
There were two commits in the old branch that weren't include.  I cherry-picked them over.